src/prefect/_internal/compatibility/async_compatible.py

```diff
@@ -0,0 +1,70 @@
+import asyncio
+import inspect
+from functools import wraps
+from typing import Any, Callable, Coroutine, Protocol, TypeVar, Union
+
+from typing_extensions import ParamSpec
+
+R = TypeVar("R")
+P = ParamSpec("P")
+
+
+class AsyncCompatible(Protocol[P, R]):
+    """Protocol for functions decorated with async_compatible."""
+
+    def __call__(
+        self, *args: P.args, **kwargs: P.kwargs
+    ) -> Union[R, Coroutine[Any, Any, R]]:
+        ...
+
+    aio: Callable[P, Coroutine[Any, Any, R]]
+
+
+def is_in_async_context() -> bool:
+    """Check if we're in an async context."""
+    try:
+        # First check if we're in a coroutine
+        if asyncio.current_task() is not None:
+            return True
+
+        # Check if we have a loop and it's running
+        loop = asyncio.get_event_loop()
+        return loop.is_running()
+    except RuntimeError:
+        return False
+
+
+def async_compatible(
```

Comment:
The `sync_compatible` name made some sense because it specified a _property_ of the callable it was attached to; in this case, I think we might want to consider naming this `async_dispatch` to highlight what it's doing more clearly.

Reply:
updated in [f9ae74b](https://github.com/PrefectHQ/prefect/pull/15813/commits/f9ae74b9fbc828d45842396faec954a90dad09a8)